left sidebar: Re-narrow after initial streams list rebuild after reload.

synicalsyntax opened the pull request:
[![https://gyazo.com/23e85037b880f37a39a5c6e07a065bf7](https://i.gyazo.com/23e85037b880f37a39a5c6e07a065bf7.gif)](https://gyazo.com/23e85037b880f37a39a5c6e07a065bf7)

We use `handle_narrow_activated` instead of `update_sidebar_for_narrow` because the latter doesn't call `scroll_stream_into_view`, resulting in wrong scroll position of the left sidebar after the streams list get rebuilt after reload. `handle_narrow_activated` also calls `update_sidebar_for_narrow`, so there should be no issues in switching to this function.

Fixes #9043.

showell commented:
Hi, @synicalsyntax, thanks for digging into this.  This is a tricky issue, because there are lots of different circumstances in which the stream list gets updated, but we want to be careful not to update the scroll in situations where the current user hasn't really done anything.  For example, suppose Cordelia is narrowed to stream A, but she starts scrolling down to page 2 looking for stream Y.  If at the same time Aaron renames stream B, we want the rename-stream event to redraw Cordelia's sidebar without jumping her scroll back up to page 1.

So, long story short, I think your fix is a bit too aggressive.  I'm not convinced we want to call `handle_narrow_activated` every time we call `update_streams_sidebar`.

We also want to be careful about names here.  The function `handle_narrow_activated` should only be called when you're activating a narrow.  We could rename it to something more general, of course, but I think it's better to just explicitly call `scroll_stream_into_view` in places where that makes sense.

timabbott commented:
I agree that calling `scroll_stream_into_view` is probably better.

synicalsyntax commented:
When a narrow is reloaded, the following things happened:
1. `subs.initialize()` was called, which in turned called `streams_list.build_stream_list()`for the **first** time
2.  `hashchange.initialize()`calls `narrow.activate()`
3. `narrow.activate()` in turn calls `streams_list.handle_narrow_activated()`
4. `streams_list.handle_narrow_activated()`calls `streams_list.scroll_stream_into_view`, which correctly scrolls the stream into view.
5. However,`message_fetch.initialize()` calls `streams_list.build_stream_list()`again for the **second** time, which triggers `streams_list.update_streams_sidebar()` (which doesn't scroll the stream into view). 
6. Step 5 occurs 2 more times, and then nothing else gets called.

Given these circumstances, where do you think it would be best to call `scroll_stream_into_view()`?

showell commented on static/js/message_fetch.js:
```diff
@@ -52,6 +52,10 @@ function process_result(data, opts) {
     stream_list.update_streams_sidebar();
     pm_list.update_private_messages();
 
+    if (!stream_list.is_scrolling) {
```
We should try to avoid sharing boolean values between two modules.  If it's necessary to share some state, you can wrap it with a function to encapsulate it better.  But in this case, it's easier to just have a function called `stream_list.maybe_scroll_narrow_into_view` that handles the `is_scrolling` case internally.  We don't want the `message_fetch` modules to be managing scrolling at all beyond a one-line callback to `stream_list`--let stream list handle that.

showell commented:
@synicalsyntax I just added a comment about keeping as much logic as possible out of `message_fetch.js`.   think @timabbott was also seeing some blueslip errors.  Not sure if it's a bug or just a case where `warn` or `info` should have been used instead of `error`--he can give you more details.

synicalsyntax commented:
@showell Alright, I updated my PR according to your feedback, and I changed the error to a warn. please let me know if there's anything else I should fix!

timabbott commented:
Merged as c3b1381c2a5bee6c0af62fa1e285f61196945627, thanks @synicalsyntax!  

The overall behavior isn't ideal here, in that things bounce around in the left sidebar (ideally, they'd just land in the perfect state immediately), but we'd need to extend `page_params` with more data to prevent that.  An it's certainly better to bounce around and land in the right place, than to bounce around and land in the wrong place.
